File: 5_8/p080.py
# one case was failed
from collections import defaultdict, deque
 
n, m = map(int, input().split())
al = [defaultdict(int) for _ in range(n)]
for i in range(m):
    a, b, c = map(int, input().split())
    al[a-1][b-1] = c
    al[b-1][a-1] = c
dist = [-1]*n
seen = [False]*n
dist[0], seen[0] = 0, True
que = deque([0])
while len(que):
    p = que.popleft()
    for k, v in al[p].items():
        cost = dist[p]+v
        if dist[k] == -1:
            dist[k] = cost
        else:
            dist[k] = min(dist[k], cost)
        if seen[k]:
            continue
        seen[k] = True
        que.append(k)
print(dist[n-1])

print('###################################')

import heapq

N, M = map(int, input().split())
A, B, C = [ None ] * M, [ None ] * M, [ None ] * M
for i in range(M):
	A[i], B[i], C[i] = map(int, input().split())

G = [ list() for i in range(N + 1) ]
for i in range(M):
	G[A[i]].append((B[i], C[i]))
	G[B[i]].append((A[i], C[i]))

max_num = 10 ** 19
dist = [ max_num ] * (N + 1)
used = [ False ] * (N + 1)
Q = list()
dist[1] = 0
heapq.heappush(Q, (0, 1))

while len(Q) >= 1:
	pos = heapq.heappop(Q)[1]
	if used[pos] == True:
		continue
	used[pos] = True
	for i in G[pos]:
		to = i[0]
		cost = dist[pos] + i[1]
		if dist[to] > cost:
			dist[to] = cost
			heapq.heappush(Q, (dist[to], to))

if dist[N] != max_num:
	print(dist[N])
else:
	print(-1)

File: 5_8/p080.go
package main

import (
	"bufio"
	"container/heap"
	"fmt"
	"os"
	"strconv"
)

var scan = bufio.NewScanner(os.Stdin)

const (
	initBufSize = 10000
	maxBufSize = 10000000
	INF = 100000000000000
)

func init() {
	buf := make([]byte, initBufSize)
	scan.Buffer(buf, maxBufSize)
	scan.Split(bufio.ScanWords)
}

type Edge struct {
	T, W int
}

type EdgeCost struct {
	v, cost, index int
}

type PriorityQueue []*EdgeCost

func (pq PriorityQueue) Len()int {return len(pq)}

// 優先度の要素
func (pq PriorityQueue) Less(i, j int) bool {
	return pq[i].cost < pq[j].cost
}

func (pq PriorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
	pq[i].index = i
	pq[j].index = j
}

func (pq *PriorityQueue) Push(x interface{}) {
	n := len(*pq)
	edge := x.(*EdgeCost)
	edge.index = n
	*pq = append(*pq, edge)
}

func (pq *PriorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	edge := old[n-1]
	old[n-1] = nil
	edge.index = -1
	*pq = old[0 : n-1]
	return edge
}

func (pq *PriorityQueue) update(edge *EdgeCost, v, cost int) {
	edge.v = v
	edge.cost = cost
	heap.Fix(pq, edge.index)
}

func main() {
	n := NumStdin()
	m := NumStdin()
	G := make([][]Edge, n)
	for i := 0; i < m; i++ {
		a := NumStdin()-1
		b := NumStdin()-1
		c := NumStdin()
		G[a] = append(G[a], Edge{b, c})
		G[b] = append(G[b], Edge{a, c})
	}

	dist := make([]int, n)
	for i := 0; i < n; i++ {
		dist[i] = INF
	}
	dist[0] = 0
	
	que := make(PriorityQueue, 1, n)
	que[0] = &EdgeCost{
		cost: 0,
		v: 0,
		index: 0,
	}
	heap.Init(&que)

	for que.Len() > 0 {
		min_v := heap.Pop(&que).(*EdgeCost)
		if dist[min_v.v] < min_v.cost {
			continue
		}
		for _, edge := range G[min_v.v] {
			cost := dist[min_v.v] + edge.W
			if dist[edge.T] > cost {
				dist[edge.T] = cost
				e := &EdgeCost{cost: dist[edge.T], v: edge.T}
				heap.Push(&que, e)
				que.update(e, e.v, e.cost)
			}
		}
	}

	if dist[n-1] < INF {
		fmt.Println(dist[n-1])
	} else {
		fmt.Println(-1)
	}
}

func NumStdin() int {
	scan.Scan()
	num, err := strconv.Atoi(scan.Text())
	if err != nil {
		panic(err)
	}
	return num
}

// func main() {
// 	n := NumStdin()
// 	m := NumStdin()
// 	G := make([][]Edge, n)
// 	for i := 0; i < m; i++ {
// 		a := NumStdin()-1
// 		b := NumStdin()-1
// 		c := NumStdin()
// 		G[a] = append(G[a], Edge{b, c})
// 		G[b] = append(G[b], Edge{a, c})
// 	}

// 	dist := make([]int, n)
// 	for i := 0; i < n; i++ {
// 		dist[i] = INF
// 	}
// 	seen := make([]bool, n)
// 	dist[0] = 0

// 	for {
// 		v := -1
// 		for u := 0; u < n; u++ {
// 			if !seen[u] && (v == -1 || dist[u] < dist[v]) {
// 				v = u
// 			}
// 		}

// 		if v == -1 {
// 			break
// 		}

// 		seen[v] = true

// 		for _, e := range G[v] {
// 			dist[e.T] = min(dist[e.T], dist[v]+e.W)
// 		}
// 	}

// 	if dist[n-1] < INF {
// 		fmt.Println(dist[n-1])
// 	} else {
// 		fmt.Println(-1)
// 	}
// }

// func min(a, b int) int {
// 	if a > b {
// 		return b
// 	}
// 	return a
// }
